### ISSUE TITLE
Reserve contract receives JS metering fees; IST fees distributed to reward pool and reserve

### ISSUE THREAD

**Issue by rowgraus** [Agoric/Agoric#4399]
This ticket covers inflows to the Reserve as well as how the Reserve covers shortfalls from vault liquidations

Inflows to the Reserve should include:
- [x] All IST generated from execution fees (gas) on the Agoric chain should flow to an IST purse controlled by the Reserve #5427
- [x] 50% of fees collected from Inter Protocol contracts go to the Reserve.  Best if this is a governable parameter. #5448

Reward distribution mechanics
- [x] Daily, 10% of total reward purse should be distributed to stakers/validators as rewards (simple smoothing function) #5427

**Comment by Tartuffo** [comment]
@warner @michaelfig Need a vBank purse for the Cosmos layer to put fees into.
That vBank purse then needs to be available in the vault contract, to cover defaults.  Needs to be setup in bootstrap.

**Comment by Tartuffo** [comment]
@dckc Dan you are the primary on this.  LMK if that is problematic.

**Comment by dtribble** [comment]
The Reserve pool now exists, so this should be unblocked.

**Comment by dckc** [comment]
> Need a vBank purse for the Cosmos layer to put fees into.

I'm looking at `packages/vats/src/vat-bank.js` and I don't see how to do this. What should I be looking at, @michaelfig ? @JimLarson ?

**Comment by michaelfig** [comment]
We will change the vbank from `getFeeCollector()` to `getRewardsCollector()` and `getReserveCollector()`, each of which is backed by a different cosmos-level module account.  `getRewardsCollector()` is backed by the `x/distr` module account, `getReserveCollector()` is backed by a new vbank-created account.

The reserve pool will get a virtual purse via bootstrap from a closely-held vbank `getReservePurse()` method, so that it can manipulate the cosmos-level module account's balance from the JS side.

**Comment by rowgraus** [comment]
Updated to include discussion on vault shortfall mechanics

**Comment by rowgraus** [comment]
Discussed in RUN Protocol engineering meeting last week.  Separating out ticket for reserve to cover shortfalls from liquidations: https://github.com/Agoric/agoric-sdk/issues/5299

**Comment by Tartuffo** [comment]
We need to make this as small as possible to handle fees in MN-1.  Maybe we don't even need this if we only use Cosmos-level feed.

**Comment by dckc** [comment]
@dtribble suggested we get more clarity on what are all the fees collected in the system and where they go. I offered to give it a whirl.

@arirubinstein I hope to get your help.

**Comment by dckc** [comment]
### Survey of fees in the system

**Solid** lines are implemented; _dotted_ lines are planned. This "JS metering" ticket seems to be about the **x/swingset** fees.

```mermaid
graph LR
    x/swingset -->|msg fee pt1| x/distribution
    x/swingset -.->|msg fee pt2| Reserve
    Vaults -->|mint fee| Collector[Reward Collector]
    Liquidation -->|penalty| Reserve
    AMM -->|swap fee| Collector
    AMM -.->|initial liquidity| Reserve
    PSM -->|swap fee| Collector
    Boost -->|mint fee| Collector
    Collector --> VBANK --> x/distribution --> stakers
    Collector -.->|overflow| Reserve
```

TODO:
 - [ ] Do fees go from the AMM / Vaults / Boost go to the reserve?
 - [ ] figure out how cosmos gas fits in.

Liquidation penalty is #5367

#### AMM, Vaults, runStake/Boost

When Inter protocol is started, in `startRewardDistributor`:

https://github.com/Agoric/agoric-sdk/blob/e570a8d103112b01a2b92dac98055e916d58fedd/packages/run-protocol/src/proposals/econ-behaviors.js#L534-L539

This uses a vat created at bootstrap: (_ISSUE: why? why not a contract in the Inter protocol?_)

https://github.com/Agoric/agoric-sdk/blob/e570a8d103112b01a2b92dac98055e916d58fedd/packages/vats/src/distributeFees.js#L42-L48

Finally, each of the 3 contracts collects fees, for example, at AMM swap

https://github.com/Agoric/agoric-sdk/blob/e570a8d103112b01a2b92dac98055e916d58fedd/packages/run-protocol/src/vpool-xyk-amm/singlePool.js#L23

creator may collect them:

https://github.com/Agoric/agoric-sdk/blob/e570a8d103112b01a2b92dac98055e916d58fedd/packages/run-protocol/src/vpool-xyk-amm/multipoolMarketMaker.js#L90-L91

#### Swingset message fees

https://github.com/Agoric/agoric-sdk/blob/e570a8d103112b01a2b92dac98055e916d58fedd/golang/cosmos/x/swingset/keeper/keeper.go#L185-L189

https://github.com/Agoric/agoric-sdk/blob/e570a8d103112b01a2b92dac98055e916d58fedd/packages/cosmic-swingset/src/sim-params.js#L35-L40

TODO:
 - [ ] confirm that they go to BLD holders via [x/distribution](https://docs.cosmos.network/master/modules/distribution/)

**Comment by Chris-Hibbert** [comment]
#5377 will add a requirement that AMM creation enforces a minimum initial liquidity. The corresponding liquidity tokens should be sent to the Reserve.

**Comment by dckc** [comment]
today's Inter meeting discussion:

 - make a contract
   - for now, leave the 50% as a contract term; postpone making it governable
